# Chapter 10. Useful Python Modules

Python has a lot of modules for doing all sorts of different tasks. 
In this chapter, we’ll look at some of the most useful ones and try some of their functions.

# Making Copies with the copy Module

The copy module contains functions for creating copies of objects. Usually, when writing a program, 
you’ll create new objects, but sometimes it’s useful to create a copy of an object, and then use that 
copy to create a new object, particularly when the process of creating an object takes several steps.

For example, suppose we have an Animal class, with an __init__ function that 
takes the parameters name, number_of_legs, and color.

In [1]:
class Animal:
        def __init__(self, species, number_of_legs, color):
            self.species = species
            self.number_of_legs = number_of_legs
            self.color = color

We could create a new object in the class Animal using the following code. 
Let’s create a pink hippogriff with six legs, called harry.

In [2]:
harry = Animal('hippogriff', 6, 'pink')

Suppose we want a herd of pink hippogriffs with six legs? 
We could repeat the previous code over and over again, or use copy, which can be found in the copy module:

In [3]:
import copy
harry = Animal('hippogriff', 6, 'pink')
harriet = copy.copy(harry)
print(harry.species)
print(harriet.species)


hippogriff
hippogriff


We can also create and copy a list of Animal objects.

In [4]:
harry = Animal('hippogriff', 6, 'pink')
carrie = Animal('chimera', 4, 'green polka dots')
billy = Animal('bogill', 0, 'paisley')
my_animals = [harry, carrie, billy]
more_animals = copy.copy(my_animals)
print(more_animals[0].species)
print(more_animals[1].species)


hippogriff
chimera


In [5]:
my_animals[0].species = 'ghoul'
print(my_animals[0].species)
print(more_animals[0].species)


ghoul
ghoul


By the same token, if we add a new animal to the first list (my_animals), 
it doesn’t appear in the copy (more_animals). 
As proof, print the length of each list after adding another animal, like this:

In [6]:
sally = Animal('sphinx', 4, 'sand')
my_animals.append(sally)
print(len(my_animals))
print(len(more_animals))


4
3


When we use deepcopy to copy my_animals, we get a new list complete with copies of all of its objects. 
As a result, changes to one of our original
Animal objects won’t affect the objects in the new list. Here’s an example:

In [8]:
more_animals = copy.deepcopy(my_animals)
my_animals[0].species = 'wyrm'
print(my_animals[0].species)
print(more_animals[0].species)


wyrm
ghoul


# Getting Random Numbers with the random Module

In [9]:
import random
print(random.randint(1, 100))


75


In [10]:
print(random.randint(100, 1000))
print(random.randint(1000, 5000))


532
3888


You might use randint to do something like create a simple (and annoying) guessing game, using a while loop, like this:

In [12]:
import random
num = random.randint(1, 100)
while True:
    print('Guess a number between 1 and 100')
    guess = input()
    i = int(guess)
    if i == num:
        print('You guessed right')
        break
    elif i < num:
                   print('Try higher')
    elif i > num:
                   print('Try lower')

Guess a number between 1 and 100
25
Try higher
Guess a number between 1 and 100
30
Try higher
Guess a number between 1 and 100
72
Try higher
Guess a number between 1 and 100
80
Try lower
Guess a number between 1 and 100
75
Try lower
Guess a number between 1 and 100
74
You guessed right


In [13]:
import random
desserts = ['ice cream', 'pancakes', 'brownies', 'cookies','candy']
print(random.choice(desserts))


brownies


# Using shuffle to Shuffle a List

In [14]:
import random
desserts = ['ice cream', 'pancakes', 'brownies', 'cookies','candy']
random.shuffle(desserts)
print(desserts)


['pancakes', 'ice cream', 'candy', 'cookies', 'brownies']


# Doing Time with the time Module

In [15]:
import time
print(time.time())


1679815166.613028


In [16]:
#Numbers of seconds since January 1, 1970

In [17]:
def lots_of_numbers(max):
        for x in range(0, max):
                print(x)

In [18]:
lots_of_numbers(1000)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

Then work out how long the function takes by modifying our program with the time module.

In [19]:
def lots_of_numbers(max):
    t1 = time.time()
    for x in range(0, max):
        print(x)
    t2 = time.time()
    print('it took %s seconds' % (t2-t1))

In [20]:
lots_of_numbers(100)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
it took 0.0041768550872802734 seconds


# Converting a Date with asctime

In [21]:
import time
print(time.asctime())

Sun Mar 26 15:22:11 2023


To call asctime with a parameter, we first create a tuple with values for the date and time. 
For example, here we assign the tuple to the variable t :

In [23]:
import time
t = (2020, 2, 23, 10, 30, 48, 6, 0, 0)
print(time.asctime(t))

Sun Feb 23 10:30:48 2020


# Taking Some Time Off with sleep

The function sleep is quite useful when you want to delay or slow down your program. 
For example, to print every second from 1 to 61, we could use the following loop:

In [26]:
for x in range(1, 61):
        print(x)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60


In [29]:
for x in range(1, 61):
        print(x)
        time.sleep(0.1)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60


# HOMEWORK

#1: Copied Cars
What will the following code print?

import copy
class Car:
    pass

car1 = Car()
car1.wheels = 4
car2 = car1
car2.wheels = 3
print(car1.wheels)
What is printed here?
car3 = copy.copy(car1)
car3.wheels = 6
print(car1.wheels)
What is printed here?

# Chapter 12. Using tkinter for Better Graphics

# Creating a Clickable Button

In [31]:
from tkinter import *
tk = Tk() #variable containing object named Tk()
btn = Button(tk, text="click me") #create a button and pass tk variable as the first parameter, click me as the second parameter
btn.pack() #tells the button to appear

In [32]:
def hello():
    print('Hello World')

In [33]:
from tkinter import *
tk = Tk() #variable containing object named Tk()
btn = Button(tk, text="click me", command=hello) #create a button and pass tk variable as the first parameter, click me as the second parameter
btn.pack() #tells the button to appear

# Creating a Canvas for Drawing

In [34]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=500, height=500)
canvas.pack()

# Drawing Lines

To draw a line on the canvas, we use pixel coordinates. 
Coordinates determine the positions of pixels on a surface. 
On a tkinter canvas, coordinates describe how far across the canvas (from left to right) 
and how far down the canvas (top to bottom) to place the pixel.

For example, since our canvas is 500 pixels wide by 500 pixels high, 
the coordinates of the bottom-right corner of the screen are (500, 500). 
To draw the line shown in the following image, we would use the starting 
coordinates (0, 0) and ending coordinates (500, 500).

In [36]:
# We specify the coordinates using the create_line function, as shown here:

from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=500, height=500)
canvas.pack()
canvas.create_line(0, 0, 500, 500)

1

# Drawing Boxes

The tkinter module makes it a lot easier to draw a square or rectangle. 
All you need to know are the coordinates for the corners. 
Here’s an example (you can close the other windows now):

In [37]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_rectangle(10, 10, 50, 50)

1

We’ll refer to these two sets of coordinates as x1, y1 and x2, y2. 
To draw a rectangle, we can increase the distance of the second corner 
from the side of the canvas (increasing the value of the x2 parameter), like this:

In [38]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_rectangle(10, 10, 50, 300)

1

# Drawing a Lot of Rectangles

How about filling the canvas with different-sized rectangles? 
We can do this by importing the module random and then 
creating a function that uses a random number for 
the coordinates at the top-left and bottom-right corners of the rectangle.

We’ll use the function provided by the random module called randrange. 
When we give this function a number, it returns a random integer 
between 0 and the number we give it. For example, calling randrange(10) 
would return a number between 0 and 9, randrange(100) 
would return a number between 0 and 99, and so on.

Here’s how we use randrange in a function. 
Create a new window by selecting File ▸ New Window, and enter the following code:

In [39]:
from tkinter import *
import random
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
def random_rectangle(width, height):
    x1 = random.randrange(width)
    y1 = random.randrange(height)
    x2 = x1 + random.randrange(width)
    y2 = y1 + random.randrange(height)
    canvas.create_rectangle(x1, y1, x2, y2)

Save the code you’ve entered (select File ▸ Save and enter a filename such as randomrect.py) and 
then select Run ▸ Run Module. Once you’ve seen the function working, 
fill the screen with rectangles by creating a loop to call random_rectangle 
a number of times. Let’s try a for loop of 100 random rectangles. 
Add the following code, save your work, and try running it again:

In [40]:
for x in range(0, 100):
    random_rectangle(400, 400)

# Setting the Color

Of course, we want to add color to our graphics. 
Let’s change the random_rectangle function to pass in a color for the 
rectangle as an additional parameter (fill_color). 
Enter this code in a new window, and when you save, call the file colorrect.py:

In [41]:
from tkinter import *
import random
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()

def random_rectangle(width, height, fill_color):
    x1 = random.randrange(width)
    y1 = random.randrange(height)
    x2 = random.randrange(x1 + random.randrange(width))
    y2 = random.randrange(y1 + random.randrange(height))
    canvas.create_rectangle(x1, y1, x2, y2, fill=fill_color)

In [42]:
random_rectangle(400, 400, 'green')
random_rectangle(400, 400, 'red')
random_rectangle(400, 400, 'blue')
random_rectangle(400, 400, 'orange')
random_rectangle(400, 400, 'yellow')
random_rectangle(400, 400, 'pink')
random_rectangle(400, 400, 'purple')
random_rectangle(400, 400, 'violet')
random_rectangle(400, 400, 'magenta')
random_rectangle(400, 400, 'cyan')

# Drawing Arcs

An arc is a segment of the circumference of a circle or another curve, but in order to draw one with tkinter, 
you need to draw it inside a rectangle using the create_arc function, with code like this:

canvas.create_arc(10, 10, 200, 100, extent=180, style=ARC)

In [43]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_arc(10, 10, 200, 100, extent=180, style=ARC)

1

In [44]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_arc(10, 10, 200, 80, extent=45, style=ARC)
canvas.create_arc(10, 80, 200, 160, extent=90, style=ARC)
canvas.create_arc(10, 160, 200, 240, extent=135, style=ARC)
canvas.create_arc(10, 240, 200, 320, extent=180, style=ARC)
canvas.create_arc(10, 320, 200, 400, extent=359, style=ARC)

5

# Drawing Polygons

# A polygon is any shape with three or more sides.

In [45]:
#Here’s how we can draw a triangle:

from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_polygon(10, 10, 100, 10, 100, 110, fill="",
outline="black")

1

We can add another irregular polygon (a shape with uneven angles or sides) using this code:

canvas.create_polygon(200, 10, 240, 30, 120, 100, 140, 120, fill="", outline="black")

# Displaying Text

In addition to drawing shapes, you can also write on the canvas using create_text. 
This function takes only two coordinates (the x and y positions of the text), 
along with a named parameter for the text to display. In the following code, we create our canvas as 
before and then display a sentence positioned at the coordinates (150, 100). Save this code as text.py.

In [46]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_text(150, 100, text='There once was a man from Toulouse,')

1

In [ ]:
The create_text function takes some other useful parameters, such as a text fill color. 
In the following code, we call the create_text function with coordinates (130, 120), 
the text we want to display, and a red fill color.

In [ ]:
canvas.create_text(130, 120, text='Who rode around on a moose.', fill='red')

In [47]:
canvas.create_text(150, 150, text='He said, "It\'s my curse,',
font=('Times', 15))
canvas.create_text(200, 200, text='But it could be worse,',
font=('Helvetica', 20))
canvas.create_text(220, 250, text='My cousin rides round',
font=('Courier', 22))
canvas.create_text(220, 300, text='on a goose."', font=('Courier', 30))

5

# Creating Basic Animation

We’ve covered how to create static drawings—pictures that don’t move. What about creating animation?

Animation is not necessarily a specialty of the tkinter module, 
but it can handle the basics. For example, we can create a 
filled triangle and then make it move across the screen using this code 
(don’t forget, select File ▸ New Window, save your work, and then run the code with Run ▸ Run Module):

In [1]:
import time
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=200)
canvas.pack()
canvas.create_polygon(10, 10, 10, 60, 50, 35)
for x in range(0, 60):
    canvas.move(1, 5, 0)
    tk.update()
    time.sleep(0.05)

To make the triangle move diagonally down the screen, we can modify this code by calling move(1, 5, 5). To try this, 
close the canvas, and create a new file (File ▸ New Window) for the following code:

import time
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_polygon(10, 10, 10, 60, 50, 35)
for x in range(0, 60):
    canvas.move(1, 5, 5)
    tk.update()
    time.sleep(0.05)

To move the triangle diagonally back up the screen to its starting position, 
use −5, −5 (add this code to the bottom of the file):

for x in range(0, 60):
    canvas.move(1, -5, -5)
    tk.update()
    time.sleep(0.05)

# Making an Object React to Something

We can make the triangle react when someone presses a key by using event bindings. 
Events are things that occur while a program is running, 
such as someone moving the mouse, pressing a key, or closing

a window. You can tell tkinter to watch for these events and then do something in response.

To begin handling events (making Python do something when an event occurs), we first create a function. 
The binding part comes when we tell tkinter that a particular function is bound (or associated) to a 
specific event; in other words, it will be automatically called by tkinter to handle that event.

For example, to make the triangle move when the enter key is pressed, we can define this function:

def movetriangle(event):
     canvas.move(1, 5, 0)

The function takes a single parameter (event), which tkinter uses to send information to the 
function about the event. We now tell tkinter that this function should be used for a particular event, 
using the bind_all function on the canvas. 
The full code now looks like this:

from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_polygon(10, 10, 10, 60, 50, 35)
def movetriangle(event):
    canvas.move(1, 5, 0)
canvas.bind_all('<KeyPress-Return>', movetriangle)

The first parameter in this function describes the event that we want tkinter to watch for. 
In this case, it’s called <KeyPress-Return>, which is a press of the enter or return key.
We tell tkinter that the movetriangle function should be called whenever this 
KeyPress event occurs. 
Run this code, click the canvas with your mouse, and then try pressing enter on your keyboard.

How about changing the direction of the triangle depending on different key presses, such as the arrow keys? 
That’s no problem. We just need to change the movetriangle function to the following:

def movetriangle(event):
    if event.keysym == 'Up':
        canvas.move(1, 0, −3)
    elif event.keysym == 'Down':
        canvas.move(1, 0, 3)
    elif event.keysym == 'Left':
        canvas.move(1, −3, 0)
    else:
        canvas.move(1, 3, 0)

The event object passed to movetriangle contains several variables. One of these variables 
is called keysym (for key symbol), which is a string that holds the value of the actual key pressed. 
The line if event.keysym == 'Up': says that if the keysym variable contains the string 'Up', 
we should call canvas.move with the parameters (1, 0, −3), as we do in the following line. If keysym contains 'Down', as in elif event.keysym == 'Down':, 
we call it with the parameters (1, 0, 3), and so on.

We then tell tkinter that the movetriangle function should be used to handle events 
from four different keys (up, down, left, and right). The following shows 
how the code looks at this point. When you enter this code, it will again be a lot 
easier if you create a new shell window by selecting File ▸ New Window. 
Before running the code, save it with a meaningful filename, such as movingtriangle.py.

In [ ]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_polygon(10, 10, 10, 60, 50, 35)
    def movetriangle(event):
        if event.keysym == 'Up':
        canvas.move(1, 0, −3)
        elif event.keysym == 'Down':
        canvas.move(1, 0, 3)
        elif event.keysym == 'Left':
        canvas.move(1, −3, 0)
        else:
            canvas.move(1, 3, 0)
            canvas.bind_all('<KeyPress-Up>', movetriangle)
            canvas.bind_all('<KeyPress-Down>', movetriangle)
            canvas.bind_all('<KeyPress-Left>', movetriangle)
            canvas.bind_all('<KeyPress-Right>', movetriangle)

On the first line of the movetriangle function, we check whether the keysym variable contains 'Up' at . 
If it does, we move the triangle upward using the move function with the parameters 1, 0, −3 at . 
The first parameter is the identifier of the triangle, the second is the amount to move to the right 
(we don’t want to move horizontally, so the value is 0), and the third is the amount to move downward (−3 pixels).

We then check whether keysym contains 'Down' at , and if so, we move the triangle down (3 pixels) at . 
The final check is whether the value is 'Left' at , and if so, we move the triangle left (−3 pixels) at . 
If none of the values are matched, the final else at  moves the triangle right at .

Now the triangle should move in the direction of the pressed arrow key.

# More Ways to Use the Identifier

Whenever we use a create_ function from the canvas, such as create_polygon or create_rectangle, 
an identifier is returned. This identifying number can be used with other canvas 
functions, as we did earlier with the move function:

In [3]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
canvas.create_polygon(10, 10, 10, 60, 50, 35)

1

In [5]:
canvas.move(1, 5, 0)

In [6]:
mytriangle = canvas.create_polygon(10, 10, 10, 60, 50, 35)
canvas.move(mytriangle, 5, 0)

Say we create a red triangle:

In [7]:
from tkinter import *
tk = Tk()
canvas = Canvas(tk, width=400, height=400)
canvas.pack()
mytriangle = canvas.create_polygon(10, 10, 10, 60, 50, 35, fill='red')

We can change the triangle to another color using itemconfig and use the 
identifier as the first parameter. The following code says, “Change the fill color of the 
object identified by the number in variable mytriangle to blue.”

In [8]:
canvas.itemconfig(mytriangle, fill='blue')

We could also give the triangle a different-colored outline, again using the identifier as the first parameter:

In [9]:
canvas.itemconfig(mytriangle, outline='red')